### PR TITLE
Add `SingleUseConsts` mir-opt pass

### PR DESCRIPTION
**Pull request by scottmcm**
The goal here is to make a pass that can be run in debug builds to simplify the common case of constants that are used just once -- that doesn't need SSA handling and avoids any potential downside of multi-use constants.  In particular, to simplify the `if T::IS_ZST` pattern that's common in the standard library.

By also handling the case of constants that are *never* actually used this fully replaces the `ConstDebugInfo` pass, since it has all the information needed to do that naturally from the traversal it needs to do anyway.

This is roughly a wash on instructions on its own (a couple regressions, a few improvements https://github.com/rust-lang/rust/pull/125910#issuecomment-2144963361), with a bunch of size improvements.  So I'd like to land it as its own PR, then do follow-ups to take more advantage of it (in the inliner, cg_ssa, etc).

r? @saethlin

### CODE REVIEW

File: compiler/rustc_mir_transform/src/single_use_consts.rs [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+            };
+
+            // We're only changing an operand, not the terminator kinds or successors
+            let basic_blocks = body.basic_blocks.as_mut_preserves_cfg();
+            let init_statement =
+                basic_blocks[init_loc.block].statements[init_loc.statement_index].replace_nop();
+            let StatementKind::Assign(place_and_rvalue) = init_statement.kind else {
+                bug!("No longer an assign?");
+            };
+            let (place, rvalue) = *place_and_rvalue;
```
Comment (saethlin) [inline compiler/rustc_mir_transform/src/single_use_consts.rs:55]: Are you deliberately avoiding using a box pattern?

File: compiler/rustc_mir_transform/src/single_use_consts.rs [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+            debug_assert_eq!(place.as_local(), Some(local));
```
Comment (saethlin) [inline compiler/rustc_mir_transform/src/single_use_consts.rs:56]: Is there a perf argument for enabling this only in debug builds? As a general rule, I prefer to have such checks either on for everyone. If a check for some absurd circumstance is going to trip, it seems more likely to trip in the wild than in CI.

**Comment by saethlin** [comment]
> this fully replaces the `ConstDebugInfo` pass,

So why haven't we deleted the pass? I know we have somewhat of a tradition of keeping MIR transforms that are off by default or broken in the tree and I think that's extremely dubious. Especially in this case. What value will someone get from `ConstDebugInfo` at `-Zmir-opt-level=3`?

File: compiler/rustc_mir_transform/src/single_use_consts.rs (hunk 1 above)
Comment (saethlin) [reply compiler/rustc_mir_transform/src/single_use_consts.rs:55]: Mm I meant if you use a box pattern you can combine these two lines. As far as I'm aware, a box pattern is the only way to do so.

**Comment by saethlin** [comment]
Thanks!
@bors r+

**Comment by Kobzol** [comment]
Wash on instruction counts, but nice binary size wins. The cycles improvements on `tt-muncher` is noise returning back to normal.

@rustbot label: +perf-regression-triaged

**Pull request merged**